<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Changjiang Xu, Delaram Pouyabahar, Veronique Voisin, and Gary D. Bader" />

<meta name="date" content="2025-04-09" />

<title>Single-cell differential expression analysis with FLASHMM</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>

<style type="text/css">
p.abstract{
text-align: center;
font-weight: bold;
}
div.abstract{
margin: auto;
width: 90%;
}
</style>


<style type="text/css">

div.csl-bib-body { }
div.csl-entry {
clear: both;
margin-bottom: 0em;
}
.hanging div.csl-entry {
margin-left:2em;
text-indent:-2em;
}
div.csl-left-margin {
min-width:2em;
float:left;
}
div.csl-right-inline {
margin-left:2em;
padding-left:1em;
}
div.csl-indent {
margin-left: 2em;
}
</style>

<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Single-cell differential expression
analysis with FLASHMM</h1>
<h4 class="author">Changjiang Xu, Delaram Pouyabahar, Veronique Voisin,
and Gary D. Bader</h4>
<h4 class="date">April 09, 2025</h4>
<div class="abstract">
<p class="abstract">Abstract</p>
FLASHMM is a package for single-cell differential expression analysis
using linear mixed-effects models (LMMs). Mixed-effects models have
become a powerful tool in single-cell studies due to their ability to
model intra-subject correlation and inter-subject variability. However,
classic LMM estimation methods face limitations in computational speed
and memory usage when applied to large-scale single-cell transcriptomics
data. The FLASHMM package provides a fast and scalable approach to
address scalability and memory efficiency in LMM fitting. This vignette
describes the methods for LMM estimation and hypothesis testing, as well
as the R functions for implementing these methods, and demonstrates the
use of the package through an example.
</div>


<div id="TOC">
<ul>
<li><a href="#methods" id="toc-methods"><span class="toc-section-number">1</span> Methods</a>
<ul>
<li><a href="#lmm-parameter-estimation-and-hypothesis-testing" id="toc-lmm-parameter-estimation-and-hypothesis-testing"><span class="toc-section-number">1.1</span> LMM parameter estimation and
hypothesis testing</a></li>
<li><a href="#fast-and-scalable-algorithm" id="toc-fast-and-scalable-algorithm"><span class="toc-section-number">1.2</span> Fast and scalable
algorithm</a></li>
<li><a href="#flashmm" id="toc-flashmm"><span class="toc-section-number">1.3</span> FLASHMM</a></li>
</ul></li>
<li><a href="#example" id="toc-example"><span class="toc-section-number">2</span> Example</a>
<ul>
<li><a href="#simulating-the-scrna-seq-data" id="toc-simulating-the-scrna-seq-data"><span class="toc-section-number">2.1</span> Simulating the scRNA-seq
data</a></li>
<li><a href="#analyzing-the-simulated-scrna-seq-data" id="toc-analyzing-the-simulated-scrna-seq-data"><span class="toc-section-number">2.2</span> Analyzing the simulated scRNA-seq
data</a>
<ul>
<li><a href="#lmm-design" id="toc-lmm-design"><span class="toc-section-number">2.2.1</span> LMM design</a></li>
<li><a href="#lmm-fitting" id="toc-lmm-fitting"><span class="toc-section-number">2.2.2</span> LMM fitting</a></li>
<li><a href="#hypothesis-testing" id="toc-hypothesis-testing"><span class="toc-section-number">2.2.3</span> Hypothesis testing</a></li>
<li><a href="#exploring-lmm-fit-and-the-de-genes" id="toc-exploring-lmm-fit-and-the-de-genes"><span class="toc-section-number">2.2.4</span> Exploring LMM fit and the DE
genes</a></li>
<li><a href="#using-ml-method" id="toc-using-ml-method"><span class="toc-section-number">2.2.5</span> Using ML method</a></li>
</ul></li>
</ul></li>
<li><a href="#remarks" id="toc-remarks"><span class="toc-section-number">3</span> Remarks</a></li>
<li><a href="#references" id="toc-references">References</a></li>
</ul>
</div>

<script>
MathJax = {
  tex: {
    tags: 'ams'
  }
};
</script>
<div id="methods" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Methods</h1>
<div id="lmm-parameter-estimation-and-hypothesis-testing" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> LMM parameter
estimation and hypothesis testing</h2>
<p>Consider the linear mixed-effects model (LMM) as expressed below
<span class="citation">(<a href="#ref-Searle2006">Searle, Casella, and
McCulloch 2006</a>)</span> <span class="math display">\[\begin{equation}\label{eq:lmm}
y = X\beta + Zb + \epsilon,
\end{equation}\]</span> where <span class="math inline">\(y\)</span> is
an <span class="math inline">\(n\times 1\)</span> vector of observed
responses, <span class="math inline">\(X\)</span> is an <span class="math inline">\(n\times p\)</span> design matrix for fixed effects
<span class="math inline">\(\beta\)</span>, <span class="math inline">\(Z\)</span> is an <span class="math inline">\(n\times q\)</span> design matrix for random
effects <span class="math inline">\(b\)</span>, and <span class="math inline">\(\epsilon\)</span> is an <span class="math inline">\(n\times 1\)</span> vector of residual errors. The
term of random effects may be a combination of various components, <span class="math display">\[
Zb = Z_1 b_1 + \cdots + Z_K b_K,
\]</span> where <span class="math inline">\(Z=[Z_1,\ldots,Z_K]\)</span>,
<span class="math inline">\(b=[b^T_1,\ldots,b^T_K]^T\)</span>, <span class="math inline">\(K\)</span> is the number of random-effect
components, and <span class="math inline">\(Z_k\)</span> is an <span class="math inline">\(n\times q_k\)</span> design matrix for the <span class="math inline">\(k\)</span>-th random-effect component. The
superscript <span class="math inline">\(T\)</span> denotes a transpose
of vector or matrix. The basic assumptions are as follows:</p>
<ol style="list-style-type: decimal">
<li>The design matrix <span class="math inline">\(X\)</span> is of full
rank, satisfying conditions of estimability for the parameters;</li>
<li>The random vectors <span class="math inline">\(b_k\)</span> and
<span class="math inline">\(\epsilon\)</span> are independent and follow
a normal distribution, <span class="math display">\[b_k \sim
N(\mathbf{0}, \sigma^2_k I_{q_k})\quad\text{and}\quad\epsilon \sim
N(\mathbf{0}, \sigma^2I_n).\]</span></li>
</ol>
<p>Here <span class="math inline">\(\mathbf{0}\)</span> is a vector or
matrix of zero elements, <span class="math inline">\(I_n\)</span> is an
<span class="math inline">\(n\times n\)</span> identity matrix, and
<span class="math inline">\(\sigma^2_k\)</span> and <span class="math inline">\(\sigma^2\)</span> are unknown parameters, called
variance components.</p>
<p><span class="citation">Hartley and Rao (<a href="#ref-HartleyRao1967">1967</a>)</span> developed the maximum
likelihood (ML) method for estimating the LMM parameters (fixed effects
and variance components). <span class="citation">Patterson and Thompson
(<a href="#ref-PattersonThompson1971">1971</a>)</span> proposed a
modified maximum likelihood procedure which partitions the data into two
mutually uncorrelated parts, one being free of the fixed effects used
for estimating variance components, called restricted maximum likelihood
(REML) method. The REML estimator is unbiased. The ML estimator of
variance components is biased in general. With variance components,
<span class="math inline">\(\theta\)</span>, estimated, the fixed
effects estimated by either ML or REML are given as follows <span class="math display">\[
\hat\beta = (X^TV_{\theta}^{-1}X )^{-1}X^TV_{\theta}^{-1}y,
\]</span> with covariance matrix <span class="math display">\[var(\hat\beta) =
(X^TV_{\theta}^{-1}X)^{-1},\]</span> where <span class="math inline">\(\theta = (\theta_0, \theta_1,\ldots,
\theta_K)\)</span>, <span class="math inline">\(\theta_0 =
\sigma^2\)</span>, <span class="math inline">\(\theta_k =
\sigma^2_k\)</span>, and <span class="math display">\[V_{\theta} =
\theta_0 I_n + \theta_1 Z_1Z_1^T + \ldots + \theta_K
Z_KZ_K^T.\]</span></p>
<p>Estimating the variance components by either ML or REML is a
numerical optimization problem. Various iterative methods based on the
log likelihood, called gradient methods, have been proposed <span class="citation">(<a href="#ref-Searle2006">Searle, Casella, and
McCulloch 2006</a>)</span>. The gradient methods are represented by the
iteration equation <span class="math display">\[\begin{equation}\label{eq:gradient}
\theta^{(i+1)} = \theta^{(i)} + \Gamma(\theta^{(i)})\frac{\partial
l(\theta^{(i)})}{\partial\theta},
\end{equation}\]</span> where <span class="math inline">\(\partial
l(\theta)/\partial\theta\)</span> is the gradient of the log likelihood
function, and <span class="math inline">\(\Gamma(\theta)\)</span> is a
modifier matrix of the gradient direction, which can be specified by
Newton–Raphson, Fisher scoring or average information. The
Newton-Raphson method uses the inverse Hessian matrix as a modifier,
while the Fisher scoring method uses the inverse of the expectation of
the Hessian matrix. The average information method uses the inverse of
the average Hessian matrix and its expectation. We use the Fisher
scoring method. The Fisher scoring method is more stable than others
because the Hessian matrix may not be positive definite but its
expectation is positive definite.</p>
<p>The hypotheses for testing fixed effects and variance components can
be respectively defined as <span class="math display">\[
H_{0, i}: \beta_i = 0 ~~\text{versus}~~H_{1,i}: \beta_i\ne 0,
\]</span> <span class="math display">\[
H_{0, k}: \theta_k \leq 0 ~~\text{versus}~~H_{1,k}: \theta_k &gt; 0,
\]</span> where <span class="math inline">\(\theta_k\)</span>, <span class="math inline">\(k=1, \ldots, K\)</span>, represent the parameters
of variance components <span class="math inline">\(\sigma^2_k\)</span>
but are allowed to be negative. The lower boundary of the parameters,
<span class="math inline">\(\theta\)</span>, can be the negative value
such that the variance-covariance matrix, <span class="math inline">\(V_{\theta}\)</span>, remains definable (positive
definite). The negative lower boundary exists and can be less than <span class="math inline">\(- \sigma^2/\lambda_{max}\)</span>, where <span class="math inline">\(\lambda_{max} &gt; 0\)</span> is the largest
singular value of <span class="math inline">\(ZZ^T\)</span>. If <span class="math inline">\(\theta_k &gt; 0\)</span>, then <span class="math inline">\(\sigma_k^2 = \theta_k\)</span> is definable and
the mixed model is well-specified. Otherwise, if <span class="math inline">\(\theta_k \le 0\)</span>, the mixed model is
miss-specified and the <span class="math inline">\(k\)</span>-th
random-effect component shouldn’t be included.</p>
<p>Allowing the parameters of variance components to take negative value
avoids the zero boundary at the null hypothesis for the variance
components. Consequently, the asymptotic normality of the maximum
likelihood estimation at the null hypothesis holds under regularity
conditions, which enables us to use z-statistics or t-statistics for
hypothesis testing of the fixed effects and variance components. The
t-statistics for fixed effects are given by <span class="math display">\[\begin{equation}
\label{eq:tcoef}
T_i = \frac{\hat\beta_i}{\sqrt{var(\hat\beta_i)}} =
\frac{\hat\beta_i}{\sqrt{var(\hat\beta)_{ii}}} ~\sim ~t(n - p).
\end{equation}\]</span> The t-statistic for a contrast, a linear
combination of the estimated fixed effects, <span class="math inline">\(c^T\hat\beta\)</span>, is <span class="math display">\[\begin{equation}
\label{eq:tcontrast}
T_c = \frac{c^T\hat\beta}{\sqrt{c^Tvar(\hat\beta) c}} \sim t(n-p).
\end{equation}\]</span> The test statistics for the parameters of
variance components are given by <span class="math display">\[\begin{equation}
\label{eq:zvarcomp}
T_{\theta_k} = \frac{\hat\theta_k}{\sqrt{[I(\hat\theta)^{-1}]_{kk}}}
\sim N(0, 1),
\end{equation}\]</span> where <span class="math inline">\(I(\theta)\)</span> is the Fisher information
matrix.</p>
</div>
<div id="fast-and-scalable-algorithm" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Fast and scalable
algorithm</h2>
<p>We developed a summary statistics based algorithm for implementing
the gradient method <span class="math inline">\(\eqref{eq:gradient}\)</span>. Let <span class="math inline">\(XX\)</span>, <span class="math inline">\(XY\)</span>, <span class="math inline">\(ZX\)</span>, <span class="math inline">\(ZY\)</span>, and <span class="math inline">\(ZZ\)</span> denote a matrix, respectively, which
define the summary statistics that are computed from cell-level data
<span class="math inline">\(X\)</span>, <span class="math inline">\(Y\)</span> and <span class="math inline">\(Z\)</span> as follows: <span class="math display">\[\begin{equation}
\label{eq:sdata}
\begin{array}{l}
XX = X^TX, ~XY = X^TY^T,\\
ZX = Z^TX, ~ZY = Z^TY^T, ~ZZ = Z^TZ,\\
Ynorm = [y_1y_1^T, \ldots, y_my_m^T],
\end{array}
\end{equation}\]</span> where <span class="math inline">\(Y = [y_1^T,
\ldots, y_m^T]^T\)</span> is a <span class="math inline">\(m\)</span>-by-<span class="math inline">\(n\)</span> matrix of gene expression profile with
each row <span class="math inline">\(y_i\)</span> corresponding to the
expression of gene <span class="math inline">\(i\)</span>, <span class="math inline">\(i=1,\ldots,m\)</span>, <span class="math inline">\(m\)</span> is the number of genes and <span class="math inline">\(n\)</span> is the number of cells (sample size).
The summary statistics can be precomputed and stored. The summary
statistics based algorithm has a complexity of <span class="math inline">\(O(m(p^3 + q^3))\)</span>, which doesn’t depend on
number of cells (sample size <span class="math inline">\(n\)</span>). In
single-cell differential expression (DE) analysis, the numbers of fixed
and random effects, <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span>, are relatively small. Therefore, the
algorithm is fast and scalable, and requires less computer memory.</p>
</div>
<div id="flashmm" class="section level2" number="1.3">
<h2><span class="header-section-number">1.3</span> FLASHMM</h2>
<p>FLASHMM package provides two functions, <em>lmm</em> and
<em>lmmfit</em>, for fitting LMM. The <em>lmm</em> function uses summary
statistics as arguments. The <em>lmmfit</em> function is a wrapper
function of <em>lmm</em>, which directly uses cell-level data and
computes the summary statistics inside the function. The <em>lmmfit</em>
function is simple to be operated but it has a limitation of memory use.
For extremely large scale data, we can precompute the summary-level data
by <span class="math inline">\(\eqref{eq:sdata}\)</span>, and then use
<em>lmm</em> function to fit LMM. FLASHMM provides <em>lmmtest</em>
function to perform statistical test on the fixed effects and the
contrasts of the fixed effects.</p>
<p>In summary, FLASHMM package provides the following main
functions.</p>
<ul>
<li><em>lmm</em>: fit LMM using summary-level data.</li>
<li><em>lmmfit</em>: fit LMM using cell-level data.</li>
<li><em>lmmtest</em>: perform statistical tests on the fixed effects and
their contrasts.</li>
<li><em>contrast.matrix</em>: construct contrast matrix combining the
fixed effects for various comparisons.</li>
<li><em>simuRNAseq</em>: simulate multi-sample multi-cell-type scRNA-seq
dataset based on a negative binomial distribution.</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="do">##Install FLASHMM from CRAN.</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co"># install.packages(&quot;FLASHMM&quot;)</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="do">##Install FLASHMM from Github.</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co"># devtools::install_github(&quot;https://github.com/Baderlab/FLASHMM&quot;)</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="do">##Load the package.</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">library</span>(FLASHMM)</span></code></pre></div>
</div>
</div>
<div id="example" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Example</h1>
<p>We use a simulated multi-sample multi-cell-type single-cell RNA-seq
(scRNA-seq) dataset to illustrate how to utilize FLASHMM to perform
single-cell differential expression analysis. In this example, we are
interested in identifying the genes differentially expressed between two
treatments (conditions) within a cell type.</p>
<div id="simulating-the-scrna-seq-data" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Simulating the
scRNA-seq data</h2>
<p>We simulate the multi-sample multi-cell-type scRNA-seq data by
simuRNAseq function in FLASHMM package. The simuRNAseq function can
generate scRNA-seq data, with or without a reference dataset. The
reference dataset contains count data and metadata. The count data is a
genes-by-cells matrix. The metadata must be a data frame containing
three columns: samples (subjects), cell-types (clusters), and treatments
(experimental conditions), and the three columns must be named ‘sam’,
‘cls’, and ‘trt’, respectively.</p>
<p>For simplicity and demonstration purposes, we use a simulated
reference dataset to generate the scRNA-seq data. First we simulate the
reference dataset.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="do">##Generate a reference dataset by simuRNAseq function.</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2502</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>refdata <span class="ot">&lt;-</span> <span class="fu">simuRNAseq</span>(<span class="at">nGenes =</span> <span class="dv">1000</span>, <span class="at">nCells =</span> <span class="dv">10000</span>)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="do">##counts</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>counts <span class="ot">&lt;-</span> refdata<span class="sc">$</span>counts</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="do">##metadata</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> refdata<span class="sc">$</span>metadata</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="fu">head</span>(metadata)</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt;       sam cls trt libsize</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt; Cell1  A5   9   A    1941</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt; Cell2  A9   9   A    1951</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co">#&gt; Cell3  A3   8   A    2005</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co">#&gt; Cell4  A6   3   A    1906</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="co">#&gt; Cell5  B8   5   B    1913</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co">#&gt; Cell6 A11   4   A    1985</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="fu">rm</span>(refdata)</span></code></pre></div>
<p>For the simulated reference dataset, we don’t need to change the
metadata column names because the columns of samples, clusters, and
treatments, are already named ‘sam’, ‘cls’, and ‘trt’, respectively.
But, for a real biological reference dataset, we need to change the
metadata column names.</p>
<p>Next we use the reference counts and metadata to simulate scRNA-seq
data that contains 100 genes and 100,000 cells from 25 samples
(subjects) and 10 clusters (cell-types) with 2 treatments. There are 10
differentially expressed genes.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="do">##Generate the scRNA-seq data by simuRNAseq function.</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2503</span>)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">simuRNAseq</span>(counts, <span class="at">metadata =</span> metadata, <span class="at">nGenes =</span> <span class="dv">100</span>, <span class="at">nCells =</span> <span class="dv">100000</span>, <span class="at">nsam =</span> <span class="dv">25</span>, </span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>       <span class="at">ncls =</span> <span class="dv">10</span>, <span class="at">ntrt =</span> <span class="dv">2</span>, <span class="at">nDEgenes =</span> <span class="dv">10</span>, <span class="at">minbeta =</span> <span class="fl">0.5</span>, <span class="at">maxbeta =</span> <span class="dv">1</span>, <span class="at">var.randomeffects =</span> <span class="fl">0.1</span>)</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="fu">str</span>(dat)</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co">#&gt; List of 5</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co">#&gt;  $ ref.mean.dispersion:&#39;data.frame&#39;: 1000 obs. of  2 variables:</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co">#&gt;   ..$ mu        : num [1:1000] 3.19 3.56 1.02 4.24 2.11 ...</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co">#&gt;   ..$ dispersion: num [1:1000] 0.814 1.476 0.791 0.829 2.319 ...</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co">#&gt;  $ metadata           :&#39;data.frame&#39;: 100000 obs. of  4 variables:</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="co">#&gt;   ..$ sam    : chr [1:100000] &quot;B10&quot; &quot;A2&quot; &quot;A10&quot; &quot;A3&quot; ...</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="co">#&gt;   ..$ cls    : chr [1:100000] &quot;8&quot; &quot;6&quot; &quot;3&quot; &quot;1&quot; ...</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="co">#&gt;   ..$ trt    : chr [1:100000] &quot;B&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; ...</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="co">#&gt;   ..$ libsize: num [1:100000] 2292 2108 2176 1938 2183 ...</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="co">#&gt;  $ counts             : num [1:100, 1:100000] 0 2 3 0 1 0 1 0 7 4 ...</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="co">#&gt;   ..- attr(*, &quot;dimnames&quot;)=List of 2</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="co">#&gt;   .. ..$ : chr [1:100] &quot;Gene260&quot; &quot;Gene724&quot; &quot;Gene774&quot; &quot;Gene616&quot; ...</span></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="co">#&gt;   .. ..$ : chr [1:100000] &quot;Cell9224&quot; &quot;Cell3950&quot; &quot;Cell530&quot; &quot;Cell2039&quot; ...</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a><span class="co">#&gt;  $ DEgenes            :&#39;data.frame&#39;: 10 obs. of  3 variables:</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a><span class="co">#&gt;   ..$ gene   : chr [1:10] &quot;Gene201&quot; &quot;Gene864&quot; &quot;Gene522&quot; &quot;Gene597&quot; ...</span></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a><span class="co">#&gt;   ..$ beta   : num [1:10] 0.977 0.819 -0.532 0.896 0.887 ...</span></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a><span class="co">#&gt;   ..$ cluster: chr [1:10] &quot;1&quot; &quot;10&quot; &quot;2&quot; &quot;2&quot; ...</span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a><span class="co">#&gt;  $ treatment          : chr &quot;B&quot;</span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a><span class="fu">rm</span>(counts, metadata)</span></code></pre></div>
<p>The simulated data contains a genes-by-cells matrix of expression
counts and a data frame of metadata consisting of 3 columns: samples
(sam), cell-types (cls) and treatments (trt).</p>
</div>
<div id="analyzing-the-simulated-scrna-seq-data" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Analyzing the
simulated scRNA-seq data</h2>
<p>We perform differential expression analysis of the simulated dataset
using FLASHMM package. We are to identify the significantly
differentially expressed genes between two treatments in a cell-type.
The analyses involve following steps: LMM design, LMM fitting and
hypothesis testing, exploring LMM fit and the differetially expressed
(DE) genes.</p>
<p><strong>LMM design</strong>: construct design matrices for fixed and
random effects described as <span class="math inline">\(\eqref{eq:lmm}\)</span>, and compute the gene
expression profile. The gene expression is taken as log-transformed
count matrix, <span class="math display">\[Y =
\log(1+\text{counts}),\]</span> in which each row corresponds to the
expression profile for a gene. The design matrix for the fixed effects
can be created by the function model.matrix, as follows <span class="math display">\[
X = model.matrix(\sim 0 + log(library.size) + cell.type +
cell.type:treatment),
\]</span> where the interaction term <span class="math inline">\(cell.type:treatment\)</span> represents the
treatment effect in a specific cell-type. <span class="math inline">\(library.size\)</span> is a normalization factor of
the scRNA-seq counts, calculated by the total sum of counts across all
genes for each cell. We consider the subjects (samples) as random
effects that reflect the intra-subject correlation and inter-subject
variability. The design matrix for the random effects is given by <span class="math display">\[
Z = model.matrix(\sim 0 + subject).
\]</span></p>
<p><strong>LMM fitting and hypothesis testing</strong>: We use either
lmm or lmmfit function to fit LMMs for all genes. With the cell-level
data matrices <span class="math inline">\(Y\)</span>, <span class="math inline">\(X\)</span> and <span class="math inline">\(Z\)</span>, the LMMs can be fit by <span class="math display">\[lmmfit(Y, X, Z, d = d),\]</span> where <span class="math inline">\(d\)</span> is a vector of the numbers of
random-effects in different components. For <span class="math inline">\(K\)</span> components of random-effects, <span class="math inline">\(d = (q_1, \ldots, q_K)\)</span>, where <span class="math inline">\(q_k\)</span> is the number of columns of the
design matrix <span class="math inline">\(Z_k\)</span> for the <span class="math inline">\(k\)</span>-th random-effect component. For a
large-scale data, the lmmfit function has a problem of computer memory
limit. In this case, it is recommended to pre-compute and store the
summary-level data: <span class="math inline">\(XX\)</span>, <span class="math inline">\(XY\)</span>, <span class="math inline">\(ZX\)</span>, <span class="math inline">\(ZY\)</span>, <span class="math inline">\(ZZ\)</span>, and <span class="math inline">\(Y_{norm}\)</span>, defined in <span class="math inline">\(\eqref{eq:sdata}\)</span>, and then use the lmm
function to fit the LMMs as follows: <span class="math display">\[lmm(XX, XY, ZX, ZY, ZZ, Ynorm = Ynorm, n = n, d =
d).\]</span> The summary-level data doesn’t depend on the sample size
<span class="math inline">\(n\)</span>. This makes lmm memory-efficient.
The default method in the lmm and lmmfit functions is restricted maximum
likelihood (REML), that is, method = ‘REML’. If you use the maximum
likelihood (ML) method to fit the LMM, set method = ‘ML’ in the lmm and
lmmfit functions.</p>
<p>In addition, the lmm and lmmfit functions perform hypothesis testing
on the fixed effects (coefficients). We can also use the lmmtest
function to conduct statistical tests on both the fixed effects and
their contrasts for various comparisons between different levels.</p>
<p>LMM fitting returns a list of estimates of LMM parameters
(coefficients and variance components), standard errors of the
estimates, covariance matrix of the coefficients (fixed effects), and
t-values and p-values for the hypothesis testing on the coefficients,
for each gene. The LMM fitting also returns the number of iterations,
the first partial derivatives of the log likelihood, and the
log-likelihood at the termination of the iterations for each gene.</p>
<p><strong>Exploring LMM fit and DE genes</strong>: We check if the LMM
fitting is convergent, and perform hypothesis testing on the variance
components of random effects. Then we identify DE genes based on
hypothesis testing p-values for the coefficients (fixed effects). If the
absolute first partial derivatives of log likelihood are all less than
the convergence tolerance, the LMM fitting converges, otherwise it
doesn’t converge. The genes for which LMM fitting doesn’t converge
should be excluded in the subsequent analysis because the estimated
coefficients for these genes are not reliable. DE genes can be
identified by adjusting p-values using the false discovery rate (FDR) or
family-wise error rate (Bonferroni correction). Additionally, we may
exclude genes with small coefficients (effect size or log-fold
change).</p>
<div id="lmm-design" class="section level3" number="2.2.1">
<h3><span class="header-section-number">2.2.1</span> LMM design</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="do">##Gene expression matrix, Y = log2(1 + counts)</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>Y <span class="ot">&lt;-</span> <span class="fu">log2</span>(<span class="dv">1</span> <span class="sc">+</span> dat<span class="sc">$</span>counts)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>dat<span class="sc">$</span>counts <span class="ot">&lt;-</span> <span class="cn">NULL</span> <span class="co">#Remove the counts.</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="do">##Design matrix for fixed effects</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> <span class="fu">log</span>(libsize) <span class="sc">+</span> cls <span class="sc">+</span> cls<span class="sc">:</span>trt, <span class="at">data =</span> dat<span class="sc">$</span>meta)</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="do">##Design matrix for random effects</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>Z <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> <span class="fu">as.factor</span>(sam), <span class="at">data =</span> dat<span class="sc">$</span>meta)</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="do">##Dimension of random effects</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">ncol</span>(Z)</span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="fu">rm</span>(dat)</span></code></pre></div>
</div>
<div id="lmm-fitting" class="section level3" number="2.2.2">
<h3><span class="header-section-number">2.2.2</span> LMM fitting</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="do">##Option 1: fit LMM by cell-level data.</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>max.iter <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>epsilon <span class="ot">&lt;-</span> <span class="fl">1e-5</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lmmfit</span>(Y, X, Z, <span class="at">d =</span> d, <span class="at">max.iter =</span> max.iter, <span class="at">epsilon =</span> epsilon)</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="do">##Option 2: fit LMM by summary-level data.</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="do">##(1) Compute the summary-level data.</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">nrow</span>(X)</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>XX <span class="ot">&lt;-</span> <span class="fu">t</span>(X)<span class="sc">%*%</span>X</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>XY <span class="ot">&lt;-</span> <span class="fu">t</span>(Y<span class="sc">%*%</span>X)</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>ZX <span class="ot">&lt;-</span> <span class="fu">t</span>(Z)<span class="sc">%*%</span>X</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>ZY <span class="ot">&lt;-</span> <span class="fu">t</span>(Y<span class="sc">%*%</span>Z)</span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>ZZ <span class="ot">&lt;-</span> <span class="fu">t</span>(Z)<span class="sc">%*%</span>Z</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>Ynorm <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(Y<span class="sc">*</span>Y)</span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="fu">rm</span>(X, Y, Z) <span class="co">#release the memory.</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="do">##(2) Fit LMM.</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a>fitss <span class="ot">&lt;-</span> <span class="fu">lmm</span>(XX, XY, ZX, ZY, ZZ, <span class="at">Ynorm =</span> Ynorm, <span class="at">n =</span> n, <span class="at">d =</span> d, </span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a>             <span class="at">max.iter =</span> max.iter, <span class="at">epsilon =</span> epsilon)</span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a></span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a><span class="fu">identical</span>(fit, fitss) <span class="co">#The two LMM fits are identical.</span></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a><span class="co">#&gt; [1] TRUE</span></span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a><span class="fu">rm</span>(fitss)</span></code></pre></div>
</div>
<div id="hypothesis-testing" class="section level3" number="2.2.3">
<h3><span class="header-section-number">2.2.3</span> Hypothesis
testing</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="do">##Testing coefficients (fixed effects)</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">lmmtest</span>(fit)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#head(test)</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="do">##The testing t-value and p-values are also provided in the LMM fit.</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="fu">range</span>(test <span class="sc">-</span> <span class="fu">cbind</span>(<span class="fu">t</span>(fit<span class="sc">$</span>coef), <span class="fu">t</span>(fit<span class="sc">$</span>t), <span class="fu">t</span>(fit<span class="sc">$</span>p))) <span class="co">#identical</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt; [1] 0 0</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#fit$t[, 1:4]</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co">#fit$p[, 1:4]</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>fit<span class="sc">$</span>coef[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co">#&gt;                   Gene260    Gene724     Gene774     Gene616</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co">#&gt; log(libsize)  0.823538186  0.9338158  0.84825642  0.47303182</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="co">#&gt; cls1         -5.033788454 -5.4098141 -5.22008950 -3.01064722</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="co">#&gt; cls10        -5.016216620 -5.4194900 -5.23064461 -2.97351422</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="co">#&gt; cls2         -5.023813654 -5.4187129 -5.23977394 -2.96924442</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="co">#&gt; cls3         -4.997721940 -5.3974915 -5.23458730 -2.97942173</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="co">#&gt; cls4         -5.020303938 -5.4443891 -5.22523214 -2.98406521</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="co">#&gt; cls5         -5.034040308 -5.4158073 -5.21912587 -2.96949879</span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="co">#&gt; cls6         -5.039858094 -5.4040391 -5.22563185 -3.01283214</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="co">#&gt; cls7         -5.032770530 -5.4027480 -5.21182988 -2.96310244</span></span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a><span class="co">#&gt; cls8         -5.027633571 -5.4037420 -5.22567389 -2.99787326</span></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a><span class="co">#&gt; cls9         -5.011484228 -5.3934569 -5.21931480 -2.97552877</span></span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a><span class="co">#&gt; cls1:trtB     0.042932556 -0.2901492 -0.05952280  0.11278682</span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="co">#&gt; cls10:trtB    0.027706615 -0.2490376 -0.03298227  0.05911209</span></span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a><span class="co">#&gt; cls2:trtB     0.043924185 -0.2725758 -0.06385726  0.06042628</span></span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a><span class="co">#&gt; cls3:trtB     0.034137125 -0.2807799 -0.03698257  0.04462700</span></span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a><span class="co">#&gt; cls4:trtB     0.048231220 -0.2272270 -0.01519083  0.08609152</span></span>
<span id="cb6-28"><a href="#cb6-28" tabindex="-1"></a><span class="co">#&gt; cls5:trtB     0.040408456 -0.2817817 -0.05896284  0.05373212</span></span>
<span id="cb6-29"><a href="#cb6-29" tabindex="-1"></a><span class="co">#&gt; cls6:trtB     0.075970031 -0.2511360 -0.03689410  0.11972056</span></span>
<span id="cb6-30"><a href="#cb6-30" tabindex="-1"></a><span class="co">#&gt; cls7:trtB     0.029742613 -0.2712344 -0.07932668  0.06190330</span></span>
<span id="cb6-31"><a href="#cb6-31" tabindex="-1"></a><span class="co">#&gt; cls8:trtB     0.021522916 -0.2660295 -0.04673630  0.07685644</span></span>
<span id="cb6-32"><a href="#cb6-32" tabindex="-1"></a><span class="co">#&gt; cls9:trtB     0.005106265 -0.3013190 -0.08311758  0.06462308</span></span></code></pre></div>
<p>The coefficients of the interaction term cls<span class="math inline">\(\ast\)</span>:trtB represent the effects of the
treatment B versus A in a cell-type (cls<span class="math inline">\(\ast\)</span>). We can make a comparison of the
treatment B versus A within all cell-types by constructing a contrast by
summing the treatment effects across cell-types as follows.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="st">&quot;BvsA&quot;</span> <span class="ot">=</span> <span class="fu">numeric</span>(<span class="fu">nrow</span>(fit<span class="sc">$</span>coef)))</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>index <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">&quot;:&quot;</span>, <span class="fu">rownames</span>(fit<span class="sc">$</span>coef))</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>contrast[index, ] <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span><span class="fu">length</span>(index)</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="do">##Test the contrast.</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">lmmtest</span>(fit, <span class="at">contrast =</span> contrast)</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="fu">head</span>(test)</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="co">#&gt;           BvsA_coef     BvsA_t     BvsA_p</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="co">#&gt; Gene260  0.03696820  0.4499201 0.65276901</span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="co">#&gt; Gene724 -0.26912699 -2.0665983 0.03877459</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="co">#&gt; Gene774 -0.05135733 -0.4228609 0.67239767</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="co">#&gt; Gene616  0.07398792  1.4665370 0.14250518</span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="co">#&gt; Gene581  0.05802063  0.7702806 0.44113530</span></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a><span class="co">#&gt; Gene591 -0.06573230 -0.8597515 0.38992809</span></span></code></pre></div>
<p>The contrast can also be constructed by contrast.matrix function as
follows.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>BvsA <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">paste0</span>(<span class="st">&quot;cls&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="st">&quot;:trtB&quot;</span>), <span class="at">collapse =</span> <span class="st">&quot;+&quot;</span>)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>BvsA <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;(&quot;</span>, BvsA, <span class="st">&quot;)/10&quot;</span>)</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>BvsA</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">#&gt; [1] &quot;(cls1:trtB+cls2:trtB+cls3:trtB+cls4:trtB+cls5:trtB+cls6:trtB+cls7:trtB+cls8:trtB+cls9:trtB+cls10:trtB)/10&quot;</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">contrast.matrix</span>(<span class="at">contrast =</span> <span class="fu">c</span>(<span class="at">BvsA =</span> BvsA), <span class="at">model.matrix.names =</span> <span class="fu">rownames</span>(fit<span class="sc">$</span>coef))</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">lmmtest</span>(fit, <span class="at">contrast =</span> contrast)</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="fu">head</span>(test)</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt;           BvsA_coef     BvsA_t     BvsA_p</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt; Gene260  0.03696820  0.4499201 0.65276901</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt; Gene724 -0.26912699 -2.0665983 0.03877459</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt; Gene774 -0.05135733 -0.4228609 0.67239767</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co">#&gt; Gene616  0.07398792  1.4665370 0.14250518</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="co">#&gt; Gene581  0.05802063  0.7702806 0.44113530</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co">#&gt; Gene591 -0.06573230 -0.8597515 0.38992809</span></span></code></pre></div>
</div>
<div id="exploring-lmm-fit-and-the-de-genes" class="section level3" number="2.2.4">
<h3><span class="header-section-number">2.2.4</span> Exploring LMM fit
and the DE genes</h3>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="do">##(1) Check the genes for which LMM fitting converges.</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>gc <span class="ot">&lt;-</span> (<span class="fu">apply</span>(<span class="fu">abs</span>(fit<span class="sc">$</span>dlogL), <span class="dv">2</span>, max) <span class="sc">&lt;</span> epsilon) </span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="fu">sum</span>(gc) </span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co">#&gt; [1] 100</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="do">##(2) Hypothesis testing for variance components:</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="do">##    H0, theta &lt;/= 0 vs H1, theta &gt; 0.</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>z <span class="ot">&lt;-</span> fit<span class="sc">$</span>theta[<span class="st">&quot;var1&quot;</span>, ]<span class="sc">/</span>fit<span class="sc">$</span>se.theta[<span class="st">&quot;var1&quot;</span>, ]</span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">pnorm</span>(z, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="fu">sum</span>(p <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="co">#&gt; [1] 100</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="do">##(3) The DE genes specific to a cell-type</span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="do">##Coefficients, t-values, and p-values for the genes specific to a cell-type.</span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a>index <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">&quot;:&quot;</span>, <span class="fu">rownames</span>(fit<span class="sc">$</span>coef))</span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a>ce <span class="ot">&lt;-</span> fit<span class="sc">$</span>coef[index, gc]</span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a>tv <span class="ot">&lt;-</span> fit<span class="sc">$</span>t[index, gc]</span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a>pv <span class="ot">&lt;-</span> fit<span class="sc">$</span>p[index, gc]</span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a>    <span class="at">gene =</span> <span class="fu">rep</span>(<span class="fu">colnames</span>(ce), <span class="fu">nrow</span>(ce)), </span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a>    <span class="at">cluster =</span> <span class="fu">rep</span>(<span class="fu">rownames</span>(ce), <span class="at">each =</span> <span class="fu">ncol</span>(ce)),</span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a>    <span class="at">coef =</span> <span class="fu">c</span>(<span class="fu">t</span>(ce)), <span class="at">t =</span> <span class="fu">c</span>(<span class="fu">t</span>(tv)), <span class="at">p =</span> <span class="fu">c</span>(<span class="fu">t</span>(pv)))</span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a></span>
<span id="cb9-25"><a href="#cb9-25" tabindex="-1"></a><span class="do">##Adjust p-values by FDR.</span></span>
<span id="cb9-26"><a href="#cb9-26" tabindex="-1"></a>out<span class="sc">$</span>FDR <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(out<span class="sc">$</span>p, <span class="at">method =</span> <span class="st">&quot;fdr&quot;</span>)</span>
<span id="cb9-27"><a href="#cb9-27" tabindex="-1"></a></span>
<span id="cb9-28"><a href="#cb9-28" tabindex="-1"></a><span class="do">##The DE genes with FDR &lt; 0.05 and abs(logFC) &gt; 0.5</span></span>
<span id="cb9-29"><a href="#cb9-29" tabindex="-1"></a>out <span class="ot">&lt;-</span> out[<span class="fu">order</span>(out<span class="sc">$</span>p), ]</span>
<span id="cb9-30"><a href="#cb9-30" tabindex="-1"></a><span class="fu">rownames</span>(out) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb9-31"><a href="#cb9-31" tabindex="-1"></a>out[(out<span class="sc">$</span>FDR <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">&amp;</span> (<span class="fu">abs</span>(out<span class="sc">$</span>coef) <span class="sc">&gt;</span> <span class="fl">0.5</span>) , ]</span>
<span id="cb9-32"><a href="#cb9-32" tabindex="-1"></a><span class="co">#&gt;      gene    cluster       coef         t            p          FDR</span></span>
<span id="cb9-33"><a href="#cb9-33" tabindex="-1"></a><span class="co">#&gt; 1 Gene456  cls4:trtB  0.6159160 10.144524 3.600651e-24 3.600651e-21</span></span>
<span id="cb9-34"><a href="#cb9-34" tabindex="-1"></a><span class="co">#&gt; 2 Gene201  cls1:trtB  0.7149779  8.625628 6.465361e-18 3.232680e-15</span></span>
<span id="cb9-35"><a href="#cb9-35" tabindex="-1"></a><span class="co">#&gt; 3 Gene864 cls10:trtB  0.8716965  7.992823 1.332776e-15 4.442588e-13</span></span>
<span id="cb9-36"><a href="#cb9-36" tabindex="-1"></a><span class="co">#&gt; 4 Gene597  cls2:trtB  0.7638176  7.737909 1.020050e-14 2.550125e-12</span></span>
<span id="cb9-37"><a href="#cb9-37" tabindex="-1"></a><span class="co">#&gt; 5 Gene982  cls8:trtB  0.9979503  7.288876 3.148478e-13 6.296956e-11</span></span>
<span id="cb9-38"><a href="#cb9-38" tabindex="-1"></a><span class="co">#&gt; 6 Gene499  cls8:trtB -0.6079898 -5.757651 8.554122e-09 1.425687e-06</span></span>
<span id="cb9-39"><a href="#cb9-39" tabindex="-1"></a><span class="co">#&gt; 7 Gene699  cls7:trtB -0.5373567 -4.919868 8.674004e-07 1.239143e-04</span></span>
<span id="cb9-40"><a href="#cb9-40" tabindex="-1"></a><span class="co">#&gt; 8 Gene632  cls4:trtB -0.5355741 -4.810256 1.509563e-06 1.886953e-04</span></span>
<span id="cb9-41"><a href="#cb9-41" tabindex="-1"></a><span class="co">#&gt; 9 Gene378  cls9:trtB -0.6419090 -4.427605 9.538617e-06 1.059846e-03</span></span>
<span id="cb9-42"><a href="#cb9-42" tabindex="-1"></a></span>
<span id="cb9-43"><a href="#cb9-43" tabindex="-1"></a><span class="do">##The true DE genes</span></span>
<span id="cb9-44"><a href="#cb9-44" tabindex="-1"></a><span class="co">#dat$DEgenes</span></span></code></pre></div>
</div>
<div id="using-ml-method" class="section level3" number="2.2.5">
<h3><span class="header-section-number">2.2.5</span> Using ML
method</h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="do">##Fitting LMM by ML method</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="co">#fit &lt;- lmmfit(Y, X, Z, d = d, method = &quot;ML&quot;, max.iter = max.iter, epsilon = epsilon)</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lmm</span>(XX, XY, ZX, ZY, ZZ, <span class="at">Ynorm =</span> Ynorm, <span class="at">n =</span> n, <span class="at">d =</span> d, <span class="at">method =</span> <span class="st">&quot;ML&quot;</span>,</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>             <span class="at">max.iter =</span> max.iter, <span class="at">epsilon =</span> epsilon)</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="do">##The DE genes specific to a cell-type</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="do">##Coefficients, t-values, and p-values</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>index <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>index <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">&quot;:&quot;</span>, <span class="fu">rownames</span>(fit<span class="sc">$</span>coef))</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>ce <span class="ot">&lt;-</span> fit<span class="sc">$</span>coef[index, gc]</span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>tv <span class="ot">&lt;-</span> fit<span class="sc">$</span>t[index, gc]</span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>pv <span class="ot">&lt;-</span> fit<span class="sc">$</span>p[index, gc]</span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a></span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a>    <span class="at">gene =</span> <span class="fu">rep</span>(<span class="fu">colnames</span>(ce), <span class="fu">nrow</span>(ce)), </span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a>    <span class="at">cluster =</span> <span class="fu">rep</span>(<span class="fu">rownames</span>(ce), <span class="at">each =</span> <span class="fu">ncol</span>(ce)),</span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a>    <span class="at">coef =</span> <span class="fu">c</span>(<span class="fu">t</span>(ce)), <span class="at">t =</span> <span class="fu">c</span>(<span class="fu">t</span>(tv)), <span class="at">p =</span> <span class="fu">c</span>(<span class="fu">t</span>(pv)))</span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a></span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a><span class="do">##Adjusting p-values by FDR</span></span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a>out<span class="sc">$</span>FDR <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(out<span class="sc">$</span>p, <span class="at">method =</span> <span class="st">&quot;fdr&quot;</span>)</span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a></span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a><span class="do">##The DE genes with FDR &lt; 0.05 and abs(logFC) &gt; 0.5</span></span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a>out <span class="ot">&lt;-</span> out[<span class="fu">order</span>(out<span class="sc">$</span>p), ]</span>
<span id="cb10-25"><a href="#cb10-25" tabindex="-1"></a><span class="fu">rownames</span>(out) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb10-26"><a href="#cb10-26" tabindex="-1"></a>out[(out<span class="sc">$</span>FDR <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">&amp;</span> (<span class="fu">abs</span>(out<span class="sc">$</span>coef) <span class="sc">&gt;</span> <span class="fl">0.5</span>) , ]</span>
<span id="cb10-27"><a href="#cb10-27" tabindex="-1"></a><span class="co">#&gt;      gene    cluster       coef         t            p          FDR</span></span>
<span id="cb10-28"><a href="#cb10-28" tabindex="-1"></a><span class="co">#&gt; 1 Gene456  cls4:trtB  0.6159049 10.542755 5.661896e-26 5.661896e-23</span></span>
<span id="cb10-29"><a href="#cb10-29" tabindex="-1"></a><span class="co">#&gt; 2 Gene201  cls1:trtB  0.7149782  8.979640 2.762145e-19 1.381073e-16</span></span>
<span id="cb10-30"><a href="#cb10-30" tabindex="-1"></a><span class="co">#&gt; 3 Gene864 cls10:trtB  0.8717021  8.322877 8.691959e-17 2.897320e-14</span></span>
<span id="cb10-31"><a href="#cb10-31" tabindex="-1"></a><span class="co">#&gt; 4 Gene597  cls2:trtB  0.7638143  8.058659 7.797845e-16 1.949461e-13</span></span>
<span id="cb10-32"><a href="#cb10-32" tabindex="-1"></a><span class="co">#&gt; 5 Gene982  cls8:trtB  0.9979550  7.589598 3.236647e-14 6.473295e-12</span></span>
<span id="cb10-33"><a href="#cb10-33" tabindex="-1"></a><span class="co">#&gt; 6 Gene499  cls8:trtB -0.6079984 -5.993079 2.066072e-09 3.443453e-07</span></span>
<span id="cb10-34"><a href="#cb10-34" tabindex="-1"></a><span class="co">#&gt; 7 Gene699  cls7:trtB -0.5373524 -5.121801 3.031907e-07 4.331296e-05</span></span>
<span id="cb10-35"><a href="#cb10-35" tabindex="-1"></a><span class="co">#&gt; 8 Gene632  cls4:trtB -0.5355670 -5.009191 5.475210e-07 6.844013e-05</span></span>
<span id="cb10-36"><a href="#cb10-36" tabindex="-1"></a><span class="co">#&gt; 9 Gene378  cls9:trtB -0.6419070 -4.612034 3.992413e-06 4.436014e-04</span></span>
<span id="cb10-37"><a href="#cb10-37" tabindex="-1"></a></span>
<span id="cb10-38"><a href="#cb10-38" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb10-39"><a href="#cb10-39" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb10-40"><a href="#cb10-40" tabindex="-1"></a><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span id="cb10-41"><a href="#cb10-41" tabindex="-1"></a><span class="co">#&gt; Platform: x86_64-apple-darwin20</span></span>
<span id="cb10-42"><a href="#cb10-42" tabindex="-1"></a><span class="co">#&gt; Running under: macOS Monterey 12.7.6</span></span>
<span id="cb10-43"><a href="#cb10-43" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-44"><a href="#cb10-44" tabindex="-1"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb10-45"><a href="#cb10-45" tabindex="-1"></a><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span id="cb10-46"><a href="#cb10-46" tabindex="-1"></a><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span id="cb10-47"><a href="#cb10-47" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-48"><a href="#cb10-48" tabindex="-1"></a><span class="co">#&gt; locale:</span></span>
<span id="cb10-49"><a href="#cb10-49" tabindex="-1"></a><span class="co">#&gt; [1] C/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span id="cb10-50"><a href="#cb10-50" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-51"><a href="#cb10-51" tabindex="-1"></a><span class="co">#&gt; time zone: America/Toronto</span></span>
<span id="cb10-52"><a href="#cb10-52" tabindex="-1"></a><span class="co">#&gt; tzcode source: internal</span></span>
<span id="cb10-53"><a href="#cb10-53" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-54"><a href="#cb10-54" tabindex="-1"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb10-55"><a href="#cb10-55" tabindex="-1"></a><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span id="cb10-56"><a href="#cb10-56" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-57"><a href="#cb10-57" tabindex="-1"></a><span class="co">#&gt; other attached packages:</span></span>
<span id="cb10-58"><a href="#cb10-58" tabindex="-1"></a><span class="co">#&gt; [1] FLASHMM_1.2.1</span></span>
<span id="cb10-59"><a href="#cb10-59" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-60"><a href="#cb10-60" tabindex="-1"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb10-61"><a href="#cb10-61" tabindex="-1"></a><span class="co">#&gt;  [1] digest_0.6.37     R6_2.6.1          fastmap_1.2.0     Matrix_1.7-3     </span></span>
<span id="cb10-62"><a href="#cb10-62" tabindex="-1"></a><span class="co">#&gt;  [5] xfun_0.51         lattice_0.22-6    cachem_1.1.0      knitr_1.50       </span></span>
<span id="cb10-63"><a href="#cb10-63" tabindex="-1"></a><span class="co">#&gt;  [9] htmltools_0.5.8.1 rmarkdown_2.29    lifecycle_1.0.4   cli_3.6.4        </span></span>
<span id="cb10-64"><a href="#cb10-64" tabindex="-1"></a><span class="co">#&gt; [13] grid_4.4.3        sass_0.4.9        jquerylib_0.1.4   compiler_4.4.3   </span></span>
<span id="cb10-65"><a href="#cb10-65" tabindex="-1"></a><span class="co">#&gt; [17] rstudioapi_0.17.1 tools_4.4.3       evaluate_1.0.3    bslib_0.9.0      </span></span>
<span id="cb10-66"><a href="#cb10-66" tabindex="-1"></a><span class="co">#&gt; [21] yaml_2.3.10       rlang_1.1.5       jsonlite_1.9.1    MASS_7.3-65</span></span></code></pre></div>
</div>
</div>
</div>
<div id="remarks" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Remarks</h1>
<p>Building design matrices for fixed and random effects is a key step
in LMM-based differential expression analysis. Including library size, a
normalization factor for scRNA-seq, as a fixed effect can help reduce
p-value inflation. If necessary, we can add principal components as
fixed effects to further mitigate unknown batch effects.</p>
<p>In differential expression analysis, modeling samples (subjects) as
random effects accounts for between-sample variability and within-sample
correlation. If samples have different effects on gene expression, we
can also model them as fixed effects. However, this can lead to
overfitting when the number of samples (subjects) is large. For
scRNA-seq data, this may also result in collinearity when the samples
are nested within an experimental condition.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="unnumbered">References</h1>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-HartleyRao1967" class="csl-entry">
Hartley, H. O., and J. N. K. Rao. 1967. <span>“Maximum-Likelihood
Estimation for the Mixed Analysis of Variance Model.”</span>
<em>Biometrika</em> 54 (1/2): 93–108. <a href="http://www.jstor.org/stable/2333854">http://www.jstor.org/stable/2333854</a>.
</div>
<div id="ref-PattersonThompson1971" class="csl-entry">
Patterson, H. D., and R. Thompson. 1971. <span>“Recovery of Inter-Block
Information When Block Sizes Are Unequal.”</span> <em>Biometrika</em> 58
(3): 545–54. <a href="https://doi.org/10.2307/2334389">https://doi.org/10.2307/2334389</a>.
</div>
<div id="ref-Searle2006" class="csl-entry">
Searle, Shayle R., George Casella, and Charles E. McCulloch. 2006.
<em>Variance Components</em>. New Jersey: John Wiley &amp; Sons, Inc.
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
